### ISSUE TITLE
Coproduct constructor does not respect covariance

### ISSUE THREAD

**Issue by atemerev** [milessabin/shapeless#728]
```scala
import shapeless._
import shapeless.ops._
import shapeless.ops.coproduct._

trait A
case object Aa extends A

trait B
case object Bb extends B

type AB = A :+: B :+: CNil

val cA = Coproduct[AB](Aa) // does not compile :(
// could not find implicit value for parameter inj: shapeless.ops.coproduct.Inject[AB, Aa.type]

val aa: A = Aa
val cA = Coproduct[AB](aa) // does compile :)
```

**Comment by milessabin** [comment]
It's the `Inject` type class rather than the constructor which is to blame here. Do you want to take a swing at a fix for this?

**Comment by atemerev** [comment]
I will try, sure. Thank you for the pointer!

**Comment by Atry** [comment]
Similar to #715

**Comment by nickvcb** [comment]
Seems to affect Options as well

```
type AB = A :+: Option[B] :+: CNil

val bb: B = Bb
val cB = Coproduct[AB](Some(bb)) // does not compile :(
```
"could not find implicit value for parameter inj: shapeless.ops.coproduct.Inject[AB,Some[B]]"

@milessabin care to share more insights please?

Context:
Trying to use for marshallers between domain model and [`skeuomorph`](https://github.com/higherkindness/skeuomorph) generated Scala class from protobuf IDL.

`protobuf` 3 no longer allows `required` so all is optional => ``skeuomorph` generates` `coproduct`s of `Option` :-(

**Comment by SemanticBeeng** [comment]
Seems to affect Options as well

```
type AB = A :+: Option[B] :+: CNil

val bb: B = Bb
val cB = Coproduct[AB](Some(bb)) // does not compile :(
```
"could not find implicit value for parameter inj: shapeless.ops.coproduct.Inject[AB,Some[B]]"

@milessabin care to share more insights please?
Is there any way to go around in latest version?

Context:
Trying to use for marshallers between domain model and [`skeuomorph`](https://github.com/higherkindness/skeuomorph) Scala class generated from protobuf IDL.

`proto3` no longer allows `required` so all is optional => ``skeuomorph` generates` `coproduct`s of `Option` :disappointed:

**Comment by milessabin** [comment]
It's fixed in #804, but not released because it's been queued waiting for a binary breaking release. I guess that's really long overdue now.
